documentation/rir.md

```diff
@@ -0,0 +1,133 @@
+# RIR
+
+## Overview
+
+RIR is a bytecode, similar to assembly or the JVM bytecode. It stands for "R Intermediate Representation".
+
+R is an interpreted language, so every line of code is executed immediately when it's parsed. However, re-parsing every single statement is slow. So, as the RIR interpreter parses code, it first converts it into "RIR bytecode", then executes it. When a call to previously-parsed code is encountered, like in a loop or function call, instead of re-parsing the code, RIR will simply run the existing bytecode. Here is a visualization of how the following might be interpreted (the `>`s denote what the interpreter is reading, note that the actual bytecode generated is more complex):
+
+```r
```

Comment:
can you shrink this example. It's too big. Maybe just put the source code and the bytecode side-by side?